bin/oneoff/backfill_data/user_cap_state_columns

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env ruby
+
+# This script migrates CAP (Child Account Policy) data from properties to separate indexed columns.
+
+require 'ruby-progressbar'
+require_relative '../../../dashboard/config/environment'
+
+options = {skip_update: false}
+OptionParser.new do |opts|
+  opts.banner = "Usage: #{File.basename(__FILE__)} [options]"
+  opts.on('-s', '--skip-update', 'Skip performing the update.') do
+    options[:skip_update] = true
+  end
+  opts.on('-h', '--help', 'Add -s to skip the actual update.') do
+    puts opts
+    exit
+  end
+end.parse!
+
+# @see https://github.com/code-dot-org/code-dot-org/pull/52244
+CAP_RELEASE_DATE = DateTime.parse('2023-06-09')
+users = User.with_deleted.where(current_sign_in_at: CAP_RELEASE_DATE..)
+
+total_successes = 0
+total_failures = 0
+
+success_status = options[:skip_update] ? 'skipped' : 'updated'
+update_progress_title = proc {"Processed[%c/%C]: |%W| #{success_status.capitalize}: #{total_successes}; Failed: #{total_failures}; %a"}
+progress_bar = ProgressBar.create(total: users.count, format: update_progress_title.call)
+
+log_data = {'updated' => {}, 'skipped' => {}, 'failed' => {}}
+log_path = CDO.dir('log', "user_cap_state_updates_#{Time.now.utc.strftime('%Y%m%dT%H%M%S')}.json")
+
+FileUtils.mkdir_p File.dirname(log_path)
+FileUtils.touch log_path
+
+log_operation = proc do |status, user, message|
+  case status
+  when 'updated'
+    total_successes += 1
+    progress_bar.log "[#{Time.now.utc.iso8601}] Successfully updated User[#{user.id}]: #{message}".green
+  when 'skipped'
+    total_successes += 1
+    progress_bar.log "[#{Time.now.utc.iso8601}] Skipped updating User[#{user.id}]: #{message}".yellow
+  when 'failed'
+    total_failures += 1
+    progress_bar.log "[#{Time.now.utc.iso8601}] Failed to update User[#{user.id}]: #{message}".red
+  end
+
+  log_data[status][user.id] = message
+  File.write log_path, JSON.pretty_generate(log_data)
```

Comment:
Maybe I'm being a little pedantic for a oneoff script, but is it going to be a problem generating this thousands of entries wide dictionary repeatedly? I could see it getting ever so slightly slower as it goes haha. Maybe it is more reasonable to just do an appended log / CSV? You're storing 'message' here as a JSON string anyway, I believe.

Reply:
https://github.com/code-dot-org/code-dot-org/pull/59962/commits/46951522ab5aeef954df90dd51dcc10d19e5f5b7